# 4.5: Data Consistency Checks

## Importing Libs

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

## Importing Dataset

In [2]:
# importing products dataset
path = r'C:\Users\shyou\OneDrive\Documents\Data Analitics Course\Achievement4\March-2023 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [5]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

## Create test dataframe

In [8]:
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b',1 , True]


In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

### Mixed-Type Data

In [10]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
df_test.dtypes

mix    object
dtype: object

In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
df_test.dtypes

mix    object
dtype: object

### Missing Values

In [16]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
df_prods.isnull()

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49693 rows x 5 columns]

In [29]:
df_nan = df_prods.loc[df_prods['product_name'].isnull() == True]

In [30]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [31]:
df_prods.shape

(49693, 5)

In [32]:
df_nan.shape

(16, 5)

In [33]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [34]:
df_prods_clean.shape

(49677, 5)

### Duplicates

In [35]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [36]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [37]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [38]:
df_prods_clean_no_dups.shape

(49672, 5)

In [39]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

In [ ]:
# Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function

In [46]:
# importing orders_wrangled dataset
df_ords= pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [47]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [50]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [51]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

##### order_id : min: 1, max: 3421083  normal
##### user_id : min: 1, max: 206209  normal
##### order_number: min: 1, max: 100  normal
##### order_day_of_week : min: 0, max: 6  normal
##### order_hour_of_day : min: 0, max: 23  normal
##### days_since_prior_order: 0, max: 30  normal






In [52]:
# Check for mixed-type data in your df_ords dataframe

In [53]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

##### There is no mixed-type data in df_ords dataframe

In [54]:
# Run a check for missing values in your df_ords dataframe

In [55]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#####  There are 206209 null rows in days_since_prior_order   column, but in fact  they are not missing values. They represent the first purchase order for each customer
##### In my opinion, they should be kept unchanged in the data frame

In [56]:
# Run a check for duplicate values in your df_ords data

In [58]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [59]:
df_ords_dups.shape

(0, 6)

##### There is not any duplicate rows in data frame

In [ ]:
# Export your final, cleaned df_ords data as “.csv” files in your “Prepared Data” folder

In [60]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))